format/nes/nes.go

```diff
@@ -0,0 +1,329 @@
+
+	// byte 7 (half)
+	consoleType := d.FieldU2("console_type", consoleTypeMapper)
+	nes20 := d.FieldU2("nes20_identifier", scalar.UintMapDescription{0: "iNES", 1: "NES 2.0"})
+
+	if nes20 == 0b10 { // NES 2.0
+		// byte 7 second half and byte 8
+		mapperMSB := d.FieldU8("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+		d.FieldU4("submapper")
+
+		// byte 9
+		prgMSB := d.FieldU4("prg_rom_size_msb")
+		chrMSB := d.FieldU4("chr_rom_size_msb")
+		prgROMSize := romSize(prgLSB, prgMSB, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, chrMSB, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		// byte 10
+		d.FieldU4("prg_ram_size", shiftSizeMapper)
+		d.FieldU4("prg_nvram_size", shiftSizeMapper)
+
+		// byte 11
+		d.FieldU4("chr_ram_size", shiftSizeMapper)
+		d.FieldU4("chr_nvram_size", shiftSizeMapper)
+
+		// byte 12
+		d.FieldU2("cpu_ppu_timing_mode", timingModeMapper)
+		d.SeekRel(6)
+
+		// byte 13
+		switch consoleType {
+		case 1:
+			d.FieldU4("vs_ppu_type", vsPPUMapper)
+			d.FieldU4("vs_hardware_type", vsHardwareMapper)
+		case 3:
+			d.FieldU4("ext_console_type", extConsoleMapper)
+			d.SeekRel(4)
+		default:
+			d.SeekRel(8)
+		}
+
+		// byte 14
+		d.FieldU2("misc_roms")
+		d.SeekRel(6)
+
+		// byte 15
+		d.FieldU6("default_exp_device")
+		d.SeekRel(2)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	} else { // iNES
+		// byte 7 (second half)
+		mapperMSB := d.FieldU4("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+
+		prgROMSize := romSize(prgLSB, 0, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, 0, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		// byte 8
+		d.FieldU8("prg_ram_size", ramSizer(8))
+
+		// byte 9 - 15
+		d.FieldStr("crap", 7, encoding.Nop)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	}
+
+	nc.nes20 = nes20 == 0b10
+
+	if trainer {
+		nc.trainerSize = 512
+	} else {
+		nc.trainerSize = 0
+	}
+
+}
+
+func decodePRGROM(d *decode.D) {
+	for !d.End() {
+		peek := d.PeekUintBits(8)
+		peekTyp := opMap[peek].Type
+		peekArgL := ArgLength(peekTyp)
+
+		if d.BitsLeft() < int64(1+peekArgL)*8 {
+			d.FieldRawLen("junk", d.BitsLeft())
+		} else {
+
+			d.FieldStruct("instruction", func(d *decode.D) {
+				op := d.FieldU8("op_code", scalar.UintFn(func(s scalar.Uint) (scalar.Uint, error) {
+					opLookup := opMap[s.Actual]
+					s.Sym = opLookup.Name
+
+					return s, nil
+				}))
+
+				typ := opMap[op].Type
+				argL := ArgLength(typ)
+				switch argL {
+				case 1:
+					d.FieldU8("args", GetArgFormatter(typ))
+				case 2:
+					d.FieldU16("args", GetArgFormatter(typ))
+				}
+			})
+		}
+	}
+}
+
+func decodeCHRROM(d *decode.D) {
+	for !d.End() {
+		d.FieldStruct("tile", func(d *decode.D) {
+			lsbs := d.PeekBytes(8)
+			d.SeekRel(64)
+			msbs := d.PeekBytes(8)
+			d.SeekRel(-64)
+			d.FieldU64("data1")
+			d.FieldU64("data2")
```

Comment:
Could this be two FieldU64 calls?
```
lsbs := d.FieldU64("data1")
msbs := d.FieldU64("data2")
```

Reply:
Maybe, if I add another loop to shift it down to 8 bytes again. "data1" is a 8x8 1bit image. I'll give it a try.

Reply:
Made a bit of a bigger rewrite of the pixel combining code. It got rid of all the bitshifting and masking. Had to do some string sliceing and reversing instead due to endianess, but no more "peek and seek" =)